### PR TITLE
Make POMs structure and dependencies order consistent

### PR DESCRIPTION
**Pull request by wendigo**
This PR adds enforcer plugin https://github.com/ferstl/pedantic-pom-enforcers

with the following configuration:

```
<compound implementation="com.github.ferstl.maven.pomenforcers.CompoundPedanticEnforcer">
   <enforcers>POM_SECTION_ORDER,MODULE_ORDER,DEPENDENCY_MANAGEMENT_ORDER,DEPENDENCY_ORDER,PLUGIN_MANAGEMENT_ORDER</enforcers>
   <pomSectionPriorities>modelVersion,parent,groupId,artifactId,version,name,description,packaging,url,licenses,inceptionYear,scm,properties,modules</pomSectionPriorities>
   <dependenciesGroupIdPriorities>io.prestosql,io.airlift</dependenciesGroupIdPriorities>
   <dependenciesOrderBy>scope,groupId,artifactId</dependenciesOrderBy>
   <dependenciesScopePriorities>compile,runtime,provided,test</dependenciesScopePriorities>
   <dependencyManagementOrderBy>groupId,artifactId</dependencyManagementOrderBy>
   <dependencyManagementGroupIdPriorities>io.prestosql,io.airlift</dependencyManagementGroupIdPriorities>
</compound>
``` 

and fixes POMs to obey new rules.

These rules can be read as follows:

 - POM sections must be in order: modelVersion, parent, groupId, artifactId, version, name, description, packaging, url, inceptionYear, licenses, scm, properties, modules

 - Dependencies are sort by scope (compile, runtime, provided, test), and then by group id (`io.prestosql` and `io.airlift` comes first and the rest follows alphabetically) and then by artifact id alphabetically.

 - Dependency management are sort by groupId and artifactId (`io.prestosql`, `io.airlift` comes first).

### CODE REVIEW

**Comment by kokosing** [comment]
Once @electrum mentioned https://github.com/ferstl/pedantic-pom-enforcers to me. I guess there are other tools like that, but we need to first establish the convention and check if we can express that convention with any tool.

**Comment by wendigo** [comment]
@kokosing that's nice. I'll look into it. I can't see any logic in the placement and order of dependencies in main pom.

**Comment by electrum** [comment]
Let's not change the order of plugins, as that can result in build changes, since plugins run in declaration order within the same phase. I like the idea of sorting the module list, but let's do that as a separate commit.

**Comment by wendigo** [comment]
@electrum I'm not changing order of plugins. But I'm changing order of dependencies to sort them out.

File: pom.xml [hunk 1]
```diff
@@ -1299,6 +1297,15 @@
                     <artifactId>maven-enforcer-plugin</artifactId>
                     <configuration>
                         <rules>
+                            <compound implementation="com.github.ferstl.maven.pomenforcers.CompoundPedanticEnforcer">
```
Comment (losipiuk) [inline pom.xml:1300]: Does this produce understandable, actinable error messages?

File: pom.xml [hunk 2]
```diff
@@ -11,12 +11,10 @@
     <groupId>io.prestosql</groupId>
     <artifactId>presto-root</artifactId>
     <version>340-SNAPSHOT</version>
-    <packaging>pom</packaging>
-
     <name>presto-root</name>
     <description>Presto</description>
+    <packaging>pom</packaging>
```
Comment (losipiuk) [inline pom.xml:16]: this change should be in different commit

File: pom.xml (hunk 1 above)
Comment (wendigo) [reply pom.xml:1300]: Yes it does. It shows you the expected order and actual order with a diff. It's a really nice tool.

Comment (wendigo) [reply pom.xml:1300]:
```
###################################################
# COMPOUND: One does not simply write a POM file! #
###################################################

Please fix these problems:

POM_SECTION_ORDER: One does not simply write a POM file!
========================================================

Your POM has to be organized this way:

Actual Order           | Required Order
-----------------------------------------------
  modelVersion         |   modelVersion
  parent               |   parent
  groupId              |   groupId
  artifactId           |   artifactId
  version              |   version
- packaging            |
  name                 |   name
  description          |   description
                       | + packaging
  url                  |   url
  inceptionYear        |   inceptionYear
  licenses             |   licenses
  scm                  |   scm
  properties           |   properties
  modules              |   modules
  dependencyManagement |   dependencyManagement
  build                |   build
  profiles             |   profiles

```

**Comment by losipiuk** [comment]
I think this is an improvement. I guess there are cases when enforced ordering is not most natural one. E.g. one may think that that it would be more readable to have `provided` and `compile` hadoop dependencies side by side. Yet still I think we should merge it. @electrum  any objections?

**Comment by wendigo** [comment]
@losipiuk I can reorganize dependencies if needed. I just need rules to follow

**Comment by losipiuk** [comment]
> @losipiuk I can reorganize dependencies if needed. I just need rules to follow

I think this is good starting point. I think for any strict rule there will be an outlier which does not match very well. Let's stay with what you proposed. 

Thanks.

File: presto-accumulo/pom.xml [hunk 3]
```diff
@@ -252,27 +115,94 @@
...
+            <artifactId>curator-framework</artifactId>
+            <version>${dep.curator.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.apache.zookeeper</groupId>
+            <artifactId>zookeeper</artifactId>
         </dependency>
 
-        <!-- used by tests but also needed transitively -->
```
Comment (electrum) [inline presto-accumulo/pom.xml:275]: Please preserve these comments

Comment (wendigo) [reply presto-accumulo/pom.xml:275]: I've restored section comments.

File: presto-accumulo/pom.xml [hunk 4]
```diff
@@ -310,7 +241,6 @@
             <scope>provided</scope>
         </dependency>
 
-        <!-- for testing -->
```
Comment (losipiuk) [inline presto-accumulo/pom.xml:313]: gone? I personnally would not cry after removing those `for testing` thingies

File: presto-atop/pom.xml [hunk 5]
```diff
@@ -53,39 +49,43 @@
...
+        </dependency>
+
         <!-- used by tests but also needed transitively -->
         <dependency>
             <groupId>io.airlift</groupId>
             <artifactId>log-manager</artifactId>
             <scope>runtime</scope>
         </dependency>
 
-        <!-- Presto SPI -->
```
Comment (losipiuk) [inline presto-atop/pom.xml:88]: gone?

File: presto-base-jdbc/pom.xml [hunk 6]
```diff
@@ -73,115 +95,92 @@
...
-        </dependency>
-
         <!-- used by tests but also needed transitively -->
         <dependency>
             <groupId>io.airlift</groupId>
             <artifactId>json</artifactId>
             <scope>runtime</scope>
         </dependency>
 
-        <!-- Presto SPI -->
```
Comment (losipiuk) [inline presto-base-jdbc/pom.xml:115]: gone?

File: presto-cli/pom.xml [hunk 7]
```diff
@@ -94,38 +114,17 @@
...
-        </dependency>
-
-        <dependency>
-            <groupId>com.fasterxml.jackson.core</groupId>
-            <artifactId>jackson-core</artifactId>
+            <artifactId>mockwebserver</artifactId>
+            <scope>test</scope>
         </dependency>
 
-        <!-- for testing -->
```
Comment (losipiuk) [inline presto-cli/pom.xml:117]: gone?

File: presto-client/pom.xml [hunk 8]
```diff
@@ -68,7 +68,6 @@
             <artifactId>okhttp-urlconnection</artifactId>
         </dependency>
 
-        <!-- for testing -->
```
Comment (losipiuk) [inline presto-client/pom.xml:71]: gone?

File: presto-druid/pom.xml [hunk 9]
```diff
@@ -33,12 +38,10 @@
...
 
         <dependency>
-            <groupId>com.fasterxml.jackson.core</groupId>
-            <artifactId>jackson-databind</artifactId>
-            <scope>compile</scope>
+            <groupId>javax.inject</groupId>
+            <artifactId>javax.inject</artifactId>
         </dependency>
 
-        <!-- Druid JDBC Connector -->
```
Comment (losipiuk) [inline presto-druid/pom.xml:41]: gone?

File: presto-druid/pom.xml [hunk 10]
```diff
@@ -52,12 +55,6 @@
             </exclusions>
         </dependency>
 
-        <dependency>
-            <groupId>javax.inject</groupId>
-            <artifactId>javax.inject</artifactId>
-        </dependency>
-
-        <!-- SPI -->
```
Comment (losipiuk) [inline presto-druid/pom.xml:60]: gone? Also typically it is `Presto SPI`

File: presto-druid/pom.xml [hunk 11]
```diff
@@ -82,13 +79,6 @@
             <scope>provided</scope>
         </dependency>
 
-        <!-- for testing -->
```
Comment (losipiuk) [inline presto-druid/pom.xml:85]: gone?

File: presto-hive/pom.xml [hunk 12]
```diff
@@ -241,20 +246,26 @@
...
+        <!-- for benchmark -->
         <dependency>
             <groupId>io.prestosql</groupId>
-            <artifactId>presto-spi</artifactId>
-            <type>test-jar</type>
+            <artifactId>presto-benchmark</artifactId>
             <scope>test</scope>
         </dependency>
 
+        <!-- for testing -->
```
Comment (losipiuk) [inline presto-hive/pom.xml:268]: this was previously by `presto-spi`

File: presto-hive-hadoop2/pom.xml [hunk 13]
```diff
@@ -67,39 +67,39 @@
 
         <!-- for testing -->
```
Comment (losipiuk) [inline presto-hive-hadoop2/pom.xml:68]: This was before testng previoulsy

File: presto-iceberg/pom.xml [hunk 14]
```diff
@@ -200,14 +200,9 @@
 
         <!-- for testing -->
```
Comment (losipiuk) [inline presto-iceberg/pom.xml:201]: this was previosly by testng

File: presto-memory/pom.xml [hunk 15]
```diff
@@ -48,19 +43,37 @@
...
+        <!-- used by tests but also needed transitively -->
         <dependency>
-            <groupId>javax.inject</groupId>
-            <artifactId>javax.inject</artifactId>
+            <groupId>io.airlift</groupId>
+            <artifactId>log</artifactId>
+            <scope>runtime</scope>
+        </dependency>
+
+        <dependency>
```
Comment (losipiuk) [inline presto-memory/pom.xml:73]: Add `<!-- used by tests but also needed transitively -->` here too.
We have no control of order and we can end up with extra dependency between the two.

File: presto-hive/pom.xml (hunk 12 above)
Comment (wendigo) [reply presto-hive/pom.xml:268]: This looks fine for me:

```
<!-- used by tests but also needed transitively -->
        <dependency>
            <groupId>io.airlift</groupId>
            <artifactId>log-manager</artifactId>
            <scope>runtime</scope>
        </dependency>

        <dependency>
            <groupId>javax.xml.bind</groupId>
            <artifactId>jaxb-api</artifactId>
            <scope>runtime</scope>
        </dependency>

        <dependency>
            <groupId>org.xerial.snappy</groupId>
            <artifactId>snappy-java</artifactId>
            <scope>runtime</scope>
        </dependency>

        <!-- Presto SPI -->
        <dependency>
            <groupId>io.prestosql</groupId>
            <artifactId>presto-spi</artifactId>
            <scope>provided</scope>
        </dependency>

        <dependency>
            <groupId>io.airlift</groupId>
            <artifactId>slice</artifactId>
            <scope>provided</scope>
        </dependency>

        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-annotations</artifactId>
            <scope>provided</scope>
        </dependency>

        <dependency>
            <groupId>org.jetbrains</groupId>
            <artifactId>annotations</artifactId>
            <scope>provided</scope>
        </dependency>

        <dependency>
            <groupId>org.openjdk.jol</groupId>
            <artifactId>jol-core</artifactId>
            <scope>provided</scope>
        </dependency>

        <!-- for benchmark -->
        <dependency>
            <groupId>io.prestosql</groupId>
            <artifactId>presto-benchmark</artifactId>
            <scope>test</scope>
        </dependency>

        <!-- for testing -->
        <dependency>
            <groupId>io.prestosql</groupId>
            <artifactId>presto-client</artifactId>
            <scope>test</scope>
        </dependency>
```

File: presto-hive-hadoop2/pom.xml (hunk 13 above)
Comment (wendigo) [reply presto-hive-hadoop2/pom.xml:68]: I guess because testng was the first test-scoped dependency. I understand this comment means "here starts test scoped dependencies"

File: presto-rcfile/pom.xml [hunk 16]
```diff
@@ -67,22 +67,9 @@
         </dependency>
 
         <dependency>
-            <groupId>io.airlift</groupId>
-            <artifactId>aircompressor</artifactId>
-            <optional>true</optional>
-        </dependency>
-
-        <!-- for testing -->
```
Comment (losipiuk) [inline presto-rcfile/pom.xml:75]: removed?

File: presto-server-rpm/pom.xml [hunk 17]
```diff
@@ -19,40 +19,39 @@
     </properties>
 
     <dependencies>
-        <!-- for testing -->
```
Comment (losipiuk) [inline presto-server-rpm/pom.xml:22]: removed?

File: presto-spi/pom.xml [hunk 18]
```diff
@@ -23,25 +23,24 @@
...
+            <artifactId>jsr305</artifactId>
+            <optional>true</optional>
         </dependency>
 
         <dependency>
             <groupId>org.openjdk.jol</groupId>
             <artifactId>jol-core</artifactId>
         </dependency>
 
-        <!-- for testing -->
```
Comment (losipiuk) [inline presto-spi/pom.xml:41]: removed?

File: presto-iceberg/pom.xml (hunk 14 above)
Comment (wendigo) [reply presto-iceberg/pom.xml:201]: I guess this comment concerns the whole section, not a single dependency.

**Comment by losipiuk** [comment]
🎉  Meged

**Comment by kokosing** [comment]
🎉 👍
